src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs

```diff
@@ -530,42 +534,57 @@ private async ValueTask SendFrameFallbackAsync(MessageOpcode opcode, bool endOfM
+                }
 
-            // Write the payload
-            if (payloadBuffer.Length > 0)
-            {
-                payloadBuffer.CopyTo(new Span<byte>(_sendBuffer, headerLength, payloadBuffer.Length));
+                // Ensure we have a _sendBuffer.
+                AllocateSendBuffer(payloadBuffer.Length + MaxMessageHeaderLength);
+                Debug.Assert(_sendBuffer != null);
 
-                // If we added a mask to the header, XOR the payload with the mask.  We do the manipulation in the send buffer so as to avoid
-                // changing the data in the caller-supplied payload buffer.
-                if (maskOffset.HasValue)
+                // Write the message header data to the buffer.
+                int headerLength;
+                int? maskOffset = null;
+                if (_isServer)
                 {
-                    ApplyMask(new Span<byte>(_sendBuffer, headerLength, payloadBuffer.Length), _sendBuffer, maskOffset.Value, 0);
+                    // The server doesn't send a mask, so the mask offset returned by WriteHeader
+                    // is actually the end of the header.
+                    headerLength = WriteHeader(opcode, _sendBuffer, payloadBuffer, endOfMessage, useMask: false, compressed: _inflater is not null);
+                }
+                else
+                {
+                    // We need to know where the mask starts so that we can use the mask to manipulate the payload data,
+                    // and we need to know the total length for sending it on the wire.
+                    maskOffset = WriteHeader(opcode, _sendBuffer, payloadBuffer, endOfMessage, useMask: true, compressed: _inflater is not null);
+                    headerLength = maskOffset.GetValueOrDefault() + MaskLength;
                 }
-            }
 
-            // Return the number of bytes in the send buffer
-            return headerLength + payloadBuffer.Length;
+                // Write the payload
+                if (payloadBuffer.Length > 0)
+                {
+                    payloadBuffer.CopyTo(new Span<byte>(_sendBuffer, headerLength, payloadBuffer.Length));
+
+                    // If we added a mask to the header, XOR the payload with the mask.  We do the manipulation in the send buffer so as to avoid
+                    // changing the data in the caller-supplied payload buffer.
+                    if (maskOffset.HasValue)
+                    {
+                        ApplyMask(new Span<byte>(_sendBuffer, headerLength, payloadBuffer.Length), _sendBuffer, maskOffset.Value, 0);
+                    }
+                }
+
+                // Return the number of bytes in the send buffer
+                return headerLength + payloadBuffer.Length;
+            }
+            finally
+            {
+                if (_deflater is not null)
+                {
+                    _deflater.ReleaseBuffer();
```

Comment:
2 pieces of feedback:
- I don't think this needs to be in a finally.
- `_deflater?.ReleaseBuffer()` is more succinct.